src/swap/useSwapQuote.ts

```diff
@@ -1,27 +1,238 @@
 import BigNumber from 'bignumber.js'
+import { TransactionRequestCIP42 } from 'node_modules/viem/_types/chains/celo/types'
 import { useRef, useState } from 'react'
 import { useAsyncCallback } from 'react-async-hook'
 import { useSelector } from 'react-redux'
+import erc20 from 'src/abis/IERC20'
+import { STATIC_GAS_PADDING } from 'src/config'
+import { useFeeCurrencies } from 'src/fees/hooks'
 import { guaranteedSwapPriceEnabledSelector } from 'src/swap/selectors'
-import { FetchQuoteResponse, Field, ParsedSwapAmount } from 'src/swap/types'
-import { TokenBalanceWithAddress } from 'src/tokens/slice'
+import { FetchQuoteResponse, Field, ParsedSwapAmount, SwapTransaction } from 'src/swap/types'
+import { TokenBalance, TokenBalanceWithAddress } from 'src/tokens/slice'
 import Logger from 'src/utils/Logger'
+import { publicClient } from 'src/viem'
+import { estimateFeesPerGas } from 'src/viem/estimateFeesPerGas'
 import networkConfig from 'src/web3/networkConfig'
 import { walletAddressSelector } from 'src/web3/selectors'
+import { Address, EstimateGasExecutionError, Hex, encodeFunctionData } from 'viem'
 
-interface ExchangeRate {
+// Apply a multiplier for the decreased swap amount to account for the
+// varying gas costs of different swap providers (or even the same swap)
+const DECREASED_SWAP_AMOUNT_GAS_COST_MULTIPLIER = 1.2
+
+export interface QuoteResult {
   toTokenAddress: string
   fromTokenAddress: string
   swapAmount: BigNumber
   price: string
   provider: string
   estimatedPriceImpact: BigNumber | null
+  preparedTransactions?: PreparedTransactionsResult
+}
+
+interface PreparedTransactionsPossible {
+  type: 'possible'
+  approveTransaction: TransactionRequestCIP42 & { amountToApprove: bigint }
+  swapTransaction: TransactionRequestCIP42
+}
+
+export interface PreparedTransactionsNeedDecreaseSwapAmountForGas {
+  type: 'need-decrease-swap-amount-for-gas'
+  maxGasCost: BigNumber
+  feeCurrency: TokenBalance
+  decreasedSwapAmount: BigNumber
+}
+
+export interface PreparedTransactionsNotEnoughBalanceForGas {
+  type: 'not-enough-balance-for-gas'
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSwapAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+function createBaseApproveTransaction(
```

Comment:
what makes something a "base" transaction?

Reply:
Yeah naming is hard. This was suggested by Kathy, just that's it's a partial transaction which doesn't have all fields we want here.
Do you have another suggestion?